state/block-executor.md

```diff
@@ -0,0 +1,50 @@
+# Block Executor
+
+## Abstract
+
+The `BlockExecutor` is a component responsible for creating, applying, and maintaining blocks and state in the system. It interacts with the mempool and the application via the [ABCI interface].
+
+## Detailed Description
+
+The `BlockExecutor` is initialized with a proposer address, `namespace ID`, `chain ID`, `mempool`, `proxyApp`, `eventBus`, and `logger`. It uses these to manage the creation and application of blocks. It also validates blocks and commits them, updating the state as necessary.
+
+- `NewBlockExecutor`: This method creates a new instance of `BlockExecutor`. It takes a proposer address, `namespace ID`, `chain ID`, `mempool`, `proxyApp`, `eventBus`, and `logger` as parameters. See [block manager] for details.
+- `InitChain`: This method initializes the chain by calling `InitChainSync` using the consensus connection to the app. It takes a `GenesisDoc` as a parameter.
+- `CreateBlock`: This method reaps transactions from the mempool and builds a block. It takes the current state, the height of the block, last header hash, and the commit as parameters.
+- `ApplyBlock`: This method applies the block to the state. It takes the current state, the block ID, and the block itself as parameters. It can return the following named errors:
+  - `ErrEmptyValSetGenerate`: returned when applying the validator changes would result in empty set.
+  - `ErrAddingValidatorToBased`: returned when adding validators to empty validator set.
+- `Validate`: This method validates the block. It takes the state and the block as parameters. In addition to the basic [block validation] rules, it applies the following validations:
+  - New block version must match state block version.
+  - New block height must match state initial height.
+  - New block height must be adjacent to state last block height.
```

Comment:
The first validation rule should contain a qualifier, as it is only applicable at genesis (`state.LastBlockHeight is 0`). Here's how the `State` struct looks like:

```
    type State struct {
	InitialHeight uint64 // should be 1, not 0, when starting from height 1

	// LastBlockHeight=0 at genesis (ie. block(H=0) does not exist)
	LastBlockHeight uint64
    }
```

The rule is:

```
	if state.LastBlockHeight <= 0 && block.Height() != state.InitialHeight {
		return errors.New("initial block height mismatch")
	}

```

It might be more accurate to reword the rule like so:

> If state is at genesis, new block height must match state initial height.

For the second validation rule

>I think this should be specific to say must be last block height + 1 since -1 is adjacent to 0 but would be invalid.

Will update 👍🏽